File: apps/web/src/lib/api/repurpose.ts
/**
 * Repurpose API — content repurposing across platforms.
 */

import { apiFetch } from "./client";

// ── Types ────────────────────────────────────────────────

export interface RepurposeRequest {
  source_id?: string;
  source_text?: string;
  source_platform: string;
  target_platforms: string[];
  brand_id?: string;
  auto_schedule?: boolean;
}

export interface RepurposedItem {
  platform: string;
  content_type: string;
  title: string;
  body: string;
  metadata: Record<string, any>;
}

export interface RepurposeResponse {
  source_platform: string;
  repurposed: RepurposedItem[];
  scheduled_items_created: number;
}

export interface PlatformInfo {
  platform: string;
  content_type: string;
  char_limit: number | null;
  description: string;
}

// ── Source platforms ──────────────────────────────────────

export const SOURCE_PLATFORMS = [
  { value: "youtube", label: "YouTube" },
  { value: "linkedin", label: "LinkedIn" },
  { value: "twitter", label: "Twitter/X" },
  { value: "tiktok", label: "TikTok" },
  { value: "instagram", label: "Instagram" },
  { value: "blog", label: "Blog" },
  { value: "email", label: "Email" },
  { value: "other", label: "Other" },
] as const;

// ── Target platforms ─────────────────────────────────────

export const TARGET_PLATFORMS = [
  { value: "linkedin", label: "LinkedIn Post" },
  { value: "twitter", label: "Twitter/X" },
  { value: "instagram", label: "Instagram Caption" },
  { value: "tiktok", label: "TikTok Script" },
  { value: "facebook", label: "Facebook Post" },
  { value: "ad_copy", label: "Ad Copy" },
  { value: "carousel", label: "Carousel" },
  { value: "email", label: "Email Snippet" },
  { value: "blog", label: "Blog Intro" },
] as const;

// ── API Methods ──────────────────────────────────────────

export const repurposeApi = {
  repurpose: (data: RepurposeRequest) =>
    apiFetch<RepurposeResponse>("/repurpose", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  getPlatforms: () => apiFetch<PlatformInfo[]>("/repurpose/platforms"),
};

File: apps/api/app/schemas/repurpose.py
"""Pydantic schemas for the Content Repurposing system."""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, field_validator


VALID_SOURCE_PLATFORMS = {
    "youtube", "linkedin", "twitter", "tiktok", "instagram", "blog", "email", "other",
}

VALID_TARGET_PLATFORMS = {
    "linkedin", "twitter", "instagram", "tiktok", "facebook",
    "ad_copy", "carousel", "email", "blog",
}


class RepurposeRequest(BaseModel):
    source_id: Optional[str] = Field(
        None, description="ID of a scheduled_item or content_asset to repurpose",
    )
    source_text: Optional[str] = Field(
        None, max_length=50000, description="Raw text to repurpose (alternative to source_id)",
    )
    source_platform: str = Field(
        ..., description="Platform the source content was originally created for",
    )
    target_platforms: List[str] = Field(
        ..., min_length=1, max_length=8, description="Platforms to repurpose into",
    )
    brand_id: Optional[str] = None
    auto_schedule: bool = Field(
        False, description="Automatically create draft schedule items for each repurposed piece",
    )

    @field_validator("source_platform")
    @classmethod
    def validate_source_platform(cls, v: str) -> str:
        if v not in VALID_SOURCE_PLATFORMS:
            raise ValueError(f"Invalid source_platform: {v}. Valid: {sorted(VALID_SOURCE_PLATFORMS)}")
        return v

    @field_validator("target_platforms")
    @classmethod
    def validate_target_platforms(cls, v: List[str]) -> List[str]:
        invalid = set(v) - VALID_TARGET_PLATFORMS
        if invalid:
            raise ValueError(f"Invalid target platforms: {sorted(invalid)}. Valid: {sorted(VALID_TARGET_PLATFORMS)}")
        return v


class RepurposedItem(BaseModel):
    platform: str
    content_type: str
    title: str
    body: str
    metadata: Dict[str, Any] = Field(default_factory=dict)


class RepurposeResponse(BaseModel):
    source_platform: str
    repurposed: List[RepurposedItem]
    scheduled_items_created: int = 0


class PlatformInfo(BaseModel):
    platform: str
    content_type: str
    char_limit: Optional[int] = None
    description: str
